### PR TITLE
Prevent integer overflow in finance_pay_interest.

### LINKED ISSUE #5635: Issue with loan interest adding money
OS: Windows 7
Version: 0.0.8
Commit/Build: 9e1711c

I noticed that the loan interest in the Alton Towers scenario (.sc4) actually adds money instead of subtracting it. I thought it could be an issue with RCT1 importing but after additional testing in a sandbox scenario (.sc6) the issue seems to be that when the loan interest get too high it becomes a positive number. 

Screenshots:
alton towers 2017-06-16 15-46-35

### PR DESCRIPTION
**Pull request by Chaosmeister**
Fixes #5635

### CODE REVIEW

File: src/openrct2/management/finance.c [hunk 1]
```diff
@@ -94,12 +94,12 @@ void finance_payment(money32 amount, rct_expenditure_type type)
  */
 void finance_pay_wages()
 {
-    rct_peep* peep;
```
Comment (Gymnasiast) [inline src/openrct2/management/finance.c]: Unrelated change, and variables are declared on top for a good reason. Please revert.

File: src/openrct2/management/finance.c [hunk 2]
```diff
@@ -125,17 +122,15 @@ void finance_pay_research()
...
-
-    // This variable uses the 64-bit type as the line below can involve multiplying very large numbers
-    // that will overflow money32 (e.g. in the Alton Towers RCT1 scenario)
-    money64 tempcost = (current_loan * 5 * current_interest) >> 14; // (5 * interest) / 2^14 is pretty close to
-
     if (gParkFlags & PARK_FLAGS_NO_MONEY)
         return;
 
-    finance_payment(tempcost, RCT_EXPENDITURE_TYPE_INTEREST);
+    const money32 loanMultiplicator = 5 * gBankLoanInterestRate;
```
Comment (Gymnasiast) [inline src/openrct2/management/finance.c]: Multiplicator is not an English word. You probably meant 'multiplier'.

File: src/openrct2/management/finance.c [hunk 3]
```diff
@@ -125,17 +122,15 @@ void finance_pay_research()
...
+    money32 currentLoanLow = (loanMultiplicator * (gBankLoan & 0x3FFF)) >> 14; // multiply first, then bitshift.
```
Comment (Gymnasiast) [inline src/openrct2/management/finance.c]: 0x3FFF is a magic number you introduced without any explanation.

Also, Coding Style says comments should go above a line, not next to it.

**Comment by Gymnasiast** [comment]
Apart from all the line notes, this solution feels like an oddball to me. I can't help thinking there are better way to solve this.

File: src/openrct2/management/finance.c (hunk 1 above)
Comment (Chaosmeister) [inline src/openrct2/management/finance.c]: what is the reason?

File: src/openrct2/management/finance.c (hunk 3 above)
Comment (Chaosmeister) [inline src/openrct2/management/finance.c]: I can clear that up and use an identifier for the number

**Comment by Chaosmeister** [comment]
@Gymnasiast I'm absolutely on your side.  
That solution is kind of ugly and mystical.

I think, the 64-bit version would be more expressive with less code and comments. It only introduces two new types that are only used there (if there are no plans on expanding to 64-bit types in the future). But that might be negligible.

File: src/openrct2/management/finance.c (hunk 1 above)
Comment (Gymnasiast) [inline src/openrct2/management/finance.c]: Convention, and keeping code organised.

Comment (Chaosmeister) [inline src/openrct2/management/finance.c]: alright!

**Comment by Gymnasiast** [comment]
This new solution looks a whole lot better. I will still have to test it though.

File: src/openrct2/management/finance.c [hunk 4]
```diff
@@ -125,17 +125,19 @@ void finance_pay_research()
  */
 void finance_pay_interest()
 {
-    money32 current_loan = gBankLoan;
-    uint8 current_interest = gBankLoanInterestRate;
+	// This variable uses the 64-bit type as the computation below can involve multiplying very large numbers
+    // that will overflow money32 if the loan is greater than (1 << 31) / (5 * current_interest_rate)
+	money64 current_loan = gBankLoan;
+    uint8 current_interest_rate = gBankLoanInterestRate;
+	money32 interest_to_pay;
```
Comment (rwjuk) [inline src/openrct2/management/finance.c]: Shouldn't this be `money64` too, as you're storing the result of multiplying a `money64` with other values in it?

Comment (Chaosmeister) [inline src/openrct2/management/finance.c]: only when
- current_loan == 0xFFFFFFFF
- gBankLoanInterestRate > 3276,8

then the result of `(current_loan * 5 * gBankLoanInterestRate) >> 14 ` would also yield a 64bit value.

Ultimately `finance_payment(interest_to_pay, RCT_EXPENDITURE_TYPE_INTEREST);` would crop the value back to money32 - without modifications.

**Comment by Gymnasiast** [comment]
Do not merge branches into yours. Ever. You are the feature branch, which will be merged into ours, not the other way round.

**Comment by Chaosmeister** [comment]
ahh I'm still learning github.

`Merge branch 'develop' into loanfix`is the autogenerated message from github since the network version was increased in previous merges and the PR displayed it a conflict.
I just increased it to 25

**Comment by Gymnasiast** [comment]
It is the autogenerated message because it's what you did. You did not "just" increase it to 25. If your PR gets behind you should rebase it, or ask one of us to do it for you.

**Comment by Chaosmeister** [comment]
I simply used the github functionality of resolving the conflict. I did not know that it would also merge the branch into my branch.  
Well... now I know.

**Comment by Gymnasiast** [comment]
Well, anyway, I rebased it for you. It can now be merged again.
